src/rhash/hash.c

```diff
@@ -834,12 +834,18 @@ static int rc_hash_zip_file(md5_state_t* md5, void* file_handle)
     uint32_t filename_len  = RC_ZIP_READ_LE16(cdir + 0x1C);
     int32_t  extra_len     = RC_ZIP_READ_LE16(cdir + 0x1E);
     int32_t  comment_len   = RC_ZIP_READ_LE16(cdir + 0x20);
+    int32_t  external_attr = RC_ZIP_READ_LE16(cdir + 0x26);
     uint64_t local_hdr_ofs = RC_ZIP_READ_LE32(cdir + 0x2A);
     cdir_entry_len = cdirhdr_size + filename_len + extra_len + comment_len;
 
     if (signature != 0x02014b50) /* expected central directory entry signature */
       break;
 
+    /* Ignore records describing a directory (we only hash file records) */
+    name = (cdir + cdirhdr_size);
+    if (name[filename_len - 1] == '/' || name[filename_len - 1] == '\\' || (external_attr & 0x10))
```

Comment:
From what little information I can find, `external_attr` is OS-specific. Is there some guarantee that 0x10 will only apply to directory records?

Reply:
This check is based on this code and comment from miniz: https://github.com/tessel/miniz/blob/master/miniz.c#L3393

But you're right, even the comment there says this isn't perfect. According to [this stackoverflow post](https://unix.stackexchange.com/a/14727) the lower bits only ever contains DOS attribute bits so this additional check for 0x10 should be fine. To attempt to be even more correct one would have to do what the ZIP handling of 7-Zip does here [ZipItem.cpp:274](https://github.com/mcmilk/7-Zip/blob/826145b/CPP/7zip/Archive/Zip/ZipItem.cpp#L274) and do the check depending on the "version made by" attribute which contains host os information.

Info-ZIP (which is the linux Zip utility, the first open source ZIP utility?) seems to set 0x10 for directories regardless of host OS. For example the unix implementation is in [unix.c:422](https://github.com/LuaDist/zip/blob/master/unix/unix.c#L422) and it [does so on all other hosts](https://github.com/search?q=repo%3ALuaDist/zip%20MSDOS_DIR_ATTR&type=code).

But in the end, like miniz says, just checking for the name ending in a slash is probably enough, it seems every ZIP utility ever made does do that for directories.

Reply:
Given that these files _should_ be zips of FAT systems, it seems like this is a reasonable expectation.

I'm just concerned about false positive matches. Though excluding a file will change the hash, so either everyone will have the same (broken) dosz file, or the user won't be able to use theirs to earn achievements.